provider/uprops/src/binary.rs

```diff
@@ -0,0 +1,97 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::upropdump_serde;
+use icu_provider::prelude::*;
+use icu_provider::iter::IterableDataProviderCore;
+use icu_uniset::provider::*;
+use icu_uniset::UnicodeSetBuilder;
+use std::borrow::Cow;
+use std::fs;
+use std::path::PathBuf;
+
+pub struct BinaryPropertiesDataProvider {
+    root_dir: PathBuf,
+}
+
+impl BinaryPropertiesDataProvider {
+    pub fn new(root_dir: PathBuf) -> Self {
+        BinaryPropertiesDataProvider { root_dir }
+    }
+    fn get_toml_as_string(&self, name: &str) -> Result<String, DataError> {
+        let mut path: PathBuf = self.root_dir.clone().join(name);
+        path.set_extension("toml");
+        fs::read_to_string(&path).map_err(DataError::new_resc_error)
+    }
+}
+
+impl<'d, 's> DataProvider<'d, 's, UnicodePropertyV1Marker> for BinaryPropertiesDataProvider {
+    fn load_payload(
+        &self,
+        req: &DataRequest,
+    ) -> Result<DataResponse<'d, 's, UnicodePropertyV1Marker>, DataError> {
+        let toml_str: String = self.get_toml_as_string(&req.resource_path.key.sub_category)?;
+        let toml_data: upropdump_serde::binary::Main =
+            toml::from_str(&toml_str).map_err(DataError::new_resc_error)?;
+
+        let mut builder = UnicodeSetBuilder::new();
+        for (start, end) in toml_data.unicode_set.data.ranges {
+            let start = std::char::from_u32(start).ok_or(DataError::InvalidPayload)?;
+            let end = std::char::from_u32(end).ok_or(DataError::InvalidPayload)?;
```

Comment:
Nit: Please make a separate `Error` type for this crate with more detailed errors, and then use `new_resc_error` to convert it to a DataError.  `DataError::InvalidPayload` is too generic to be helpful.